[Refactor] API 호출 오류 핸들링 개선

PeraSite opened the pull request:
## Summary
ApiResult sealed class와 retrofit의 CallAdapter를 활용해 오류에 대응하는 새로운 구조로 변경합니다

## Describe your changes
- Service: BaseResponse<T> 를 ApiResult<T>로 변경
- Repository: Flow<BaseResponse<T>를 T?로 변경
  - Flow 완전 제거. ViewModel과 UseCase에서 로딩, 오류 상태 대응
  - Void 반환을 Success했는지 여부를 의미하는 Boolean으로 변경
  - 실패했다면 null을, 성공했다면 T를 반환

- ViewModel, UseCase:
  - Repository의 반환 값이 T?로 변함에 따라 모든 use-case와 repository 호출 코드 개선

## Issue

- Resolves #295

## To reviewers

- ~~local.properties 에서 백엔드 서버 도메인을 임의로 변경하고 앱을 켜보면
로그인이 되어있던 경우) 토큰 값이 유효한지 확인하지 못해 로그인 창으로 이동 > 로그인 시도하면 "로그인이 실패했습니다"
로그인이 안되어있던 경우) 로그인 창에서 로그인 시도하면 "로그인이 실패했습니다"
앱이 죽지는 않지만 ApiResult.NetworkError인 경우에는 어떻게 할 것인지 대응이 필요해요!
제 생각에 서버가 아예 죽었을 때 서버 오류 스플래시로 이동하는 걸 구현하는 제일 깔끔한 구조는 역시 health check용 use-case를 만들고, IntroActivity에서 + 정기적으로 실행하는거에요. 
이 use-case는 ApiResult를 매핑하지 않고 NetworkError를 그대로 반환해서 Presentation 레이어에서 ErrorActivity로 이동시키기만 하면 될 것 같습니다.~~

![image](https://github.com/user-attachments/assets/75cdfae5-033d-4c2e-82b7-c4bfa173c96a)
EDIT: 20251010 ServerErrorActivity 추가했습니다!

- Repository에서 ApiResult를 완전히 가리고 결과값만 던져주는 구조와, ApiResult의 data만 도메인 객체로 바꾸고 반환해주는 구조 중 무엇이 더 좋을까요? 호출하는 입장에서 '어떤 오류'가 발생했는가를 상세히 알아야할 필요가 있는지에 따라 결정하면 될 것 같습니다.

- Presentation 레이어에서 Loading 상태 변경 > API 호출 > 결과의 null 여부에 따라 Error 상태, Success 상태 변경하고 있는 구조로 변경했습니다.
그러나 상당히 반복적인 부분이 많아서 UiState를 모두 사용하게 리팩토링한 후, 자동으로 Loading과 Error 상태를 관리해주는 유틸 함수를 만들어서 쓰면 좋을 것 같아요.
```kotlin
    fun <T> networkRequest(
        uiState: MutableStateFlow<UiState<T>>,
        request: suspend () -> T,
        onSuccess: (T) -> Unit,
        onError: (Throwable) -> Unit = {},
    ) {
        viewModelScope.launch {
            uiState.value = UiState.Loading
            try {
                val result = request()
                uiState.value = UiState.Success(result)
                onSuccess(result)
            } catch (e: Exception) {
                uiState.value = UiState.Error
                onError(e)
            }
        }
    }
```

PeraSite committed:
feat: ApiResult 추가

PeraSite committed:
refactor: Repository 시그니처 정리

PeraSite committed:
feat: RepositoryImpl ApiResult 사용

PeraSite committed:
feat: Service ApiResult 사용

PeraSite committed:
feat: Presentation 레이어 ApiResult 적용

PeraSite committed:
feat: ApiResult 처리 CallAdapter 추가

PeraSite committed:
fix: getFixMenu만 크래시가 발생하던 문제 수정

PeraSite committed:
refactor: ArrayList를 List로 정리

PeraSite committed:
chore: BaseResponseConverter 삭제

PeraSite committed:
fix: 런타임 발생 오류 해결

PeraSite committed:
chore: 로그인 실패 줄바꿈 삭제

PeraSite committed:
fix: early return 추가

PeraSite committed:
refactor: 로그아웃 로직 병합

PeraSite committed:
chore: List로 타입 통일화

PeraSite committed:
feat: NetworkError 대응 Activity 추가

PeraSite committed:
feat: HealthCheck 모듈 추가

PeraSite committed:
fix: ApiResult의 함수를 외부 확장 함수로 변경

PeraSite committed:
feat: NetworkError 발생 시 ServerErrorActivity 이동 추가

PeraSite committed:
refactor: IntroActivity 함수화

PeraSite committed:
refactor: HealthCheck use-case가 boolean을 사용하게 수정, autoLogin 로직 리팩토링

PeraSite committed:
refactor: nested it 이름 변경

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+        else -> emptyList()
+    }
+
+fun ApiResult<Unit>.isSuccess(): Boolean = this is ApiResult.Success
```
단순 성공 확인용

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+        is UnknownError -> UnknownError(exception)
+    }
+
+    data class Success<T : Any>(val data: T) : ApiResult<T>
```
response가 있는 api가 성공했을 때

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+
+fun ApiResult<Unit>.isSuccess(): Boolean = this is ApiResult.Success
+
+fun <T> ApiResult<T>.orElse(default: T): T = when (this) {
```
실패 시 기본값

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+
+    data class Success<T : Any>(val data: T) : ApiResult<T>
+
+    data class Failure(
```
서버에서 에러코드 반환

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+        val message: String?
+    ) : ApiResult<Nothing>
+
+    data class NetworkError(
```
네트워크 문제

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+        val exception: IOException
+    ) : ApiResult<Nothing>
+
+    data class UnknownError(
```
알 수 없는 오류

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+
+sealed interface ApiResult<out T> {
+    fun <R : Any> map(transform: (T) -> R): ApiResult<R> = when (this) {
+        is Success -> Success(transform(data))
```
성공일 때 데이터 변환

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+
+}
+
+fun <TElement, TList : List<TElement>> ApiResult<TList>.orEmptyList(): List<TElement> =
```
리스트 응답용 헬퍼 함수, 실패 시 빈리스트 반환

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+
+import java.io.IOException
+
+sealed interface ApiResult<out T> {
```
위에 남긴 함수에 대한 설명이 제가 제대로 이해한 게 맞는지 확인 부탁드려요!
그리고 각 함수의 역할에 대해 주석이 있으면 좋을 것 같아요

PeraSite replied:
확인 후 주석 작성 완료했습니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/presentation/base/BaseActivity.kt:
```diff
@@ -176,12 +189,40 @@ abstract class BaseActivity<B : ViewBinding>(
+
+    open fun shouldCheckServerHealth(): Boolean {
+        return true
+    }
```
스타일 차이지만 요렇게 하면 뭔가 깔끔해서 기분이 좋네유
```kotlin
open fun shouldCheckServerHealth(): Boolean = true
```

PeraSite replied:
변경하려고 보니 아래 ScreenId도 같은 개념이라 수정했습니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/repository/HealthCheckRepositoryImpl.kt:
```diff
@@ -0,0 +1,18 @@
+) : HealthCheckRepository {
+    override suspend fun checkServerHealth(): Boolean =
+        when (val result = healthCheckService.checkHealth()) {
+            is ApiResult.NetworkError -> false
```
어차피 else 문 때문에 Success가 아닌 상태는 모두 false가 될텐데 네트워크 에러만 명시적으로 써두신 이유가 있을까요?

PeraSite replied:
그러네요! HealthCheckService.checkHealth 함수의 반환 타입이 BaseResponse가 아니라 키자마자 오류가 발생하길래, 반환 타입 몰라도 일단 문제가 없었으면 넘어가게끔 Unit으로 반환 타입을 변경했습니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/repository/PartnershipRepositoryImpl.kt:
```diff
@@ -13,24 +15,15 @@ class PartnershipRepositoryImpl @Inject constructor(
-            .result
-            ?.map { it.toDomain() } ?: emptyList()
-    }
+    override suspend fun getAllPartnerships(): List<Partnership> =
```
기존에 있던 주석들은 살려주세요!!!

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/repository/PartnershipRepositoryImpl.kt:
```diff
@@ -13,24 +15,15 @@ class PartnershipRepositoryImpl @Inject constructor(
-    }
+    override suspend fun getAllPartnerships(): List<Partnership> =
+        partnershipService.getAllPartnerships()
+            .map { list -> list.map { partnershipResponse -> partnershipResponse.toDomain() } }
```
```kotlin
.map { list -> list.map { it.toDomain() } }
```
이렇게도 될텐데 partnershipResponse을 쓰신 이유가 있을까요? 밑에는 it을 쓰셨길래!

PeraSite replied:
오잉 그러네요 it으로 수정하겠습니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,46 @@
+    else -> default
+}
+
+fun <T> ApiResult<T>.orNull(): T? = when (this) {
```
실패 시 null반환

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/repository/UserRepositoryImpl.kt:
```diff
@@ -2,56 +2,51 @@ package com.eatssu.android.data.repository
-        userService.changeNickname(body)
-    }
-
+    override suspend fun updateUserName(body: ChangeNicknameRequest): Boolean =
+        userService.changeNickname(body).isSuccess()
 
+    override suspend fun checkUserNameValidation(nickname: String): Boolean =
+        userService.checkNickname(nickname).orElse(false)
```
isSuccess를 쓰는 상황과 .orElse(false)를 쓰는 상황이 정확히 나뉘는 기준이 있을까요?

PeraSite replied:
updateUserName는 Response 형태가 Unit이라 BaseResponse에서 성공했는지, HTTP Response Code가 어떻게 되는지만 확인하면 되는데,
checkNickname은 Response 형태가 Boolean이라서 BaseResponse에도 성공했고, HTTP Response도 200인데 result.data 타입이 false가 될 수 있는 Boolean이길래 이렇게 작성했습니다!

실제 서버 코드를 확인해본 것은 아니지만 닉네임이 중복인게(=result.data가 false인게) 오류는 아니니까 정상적으로 반환하는 것 같아요.
즉 네트워크 요청은 Success 상태지만 닉네임이 중복이라 값이 false인 경우를 처리하는 목적입니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/repository/UserRepositoryImpl.kt:
```diff
@@ -2,56 +2,51 @@ package com.eatssu.android.data.repository
-            emit(userService.getMyReviews())
-        }
+    override suspend fun getUserNickName(): String =
+        userService.getMyInfo().map { it.nickname ?: "" }.orNull() ?: ""
```
orNull 함수를 써서 null로 받고 다시 default 값을 넣어주는 이유가 있을까요?

PeraSite replied:
map은 Success했을 때만 매핑하는 것이라 다른 오류 상태인 경우 처리까지 해주어야 ApiResult를 Unwrap할 수 있어서 그랬습니다!

정확히는 MyNickNameResponse.nickname이 String? 타입이라 Success 했을 때도 String?에 대한 기본 값을 제공해주어야 하고
실패했을 때도 기본 값을 제공해주어야 하니 두 가지 케이스를 작성한건데...

생각해보니 두 케이스 전부 String?을 반환하고 한번의 elvis operator로 기본값을 집어넣을 수 있는 것 같아 수정했습니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/data/repository/UserRepositoryImpl.kt:
```diff
@@ -2,56 +2,51 @@ package com.eatssu.android.data.repository
-        return userService.signOut().result ?: false
-    }
+    override suspend fun signOut(): Boolean =
+        userService.signOut().orElse(false)
```
여기도 .isSuccess() 랑 차이가 궁금합니다

PeraSite replied:
위 케이스와 마찬가지로 Success인데 result.data가 false인 케이스가 가능하게 API 명세가 Boolean을 반환해서 이렇게 대응했습니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/domain/repository/ReportRepository.kt:
```diff
@@ -1,13 +1,10 @@
 package com.eatssu.android.domain.repository
 
 import com.eatssu.android.data.dto.request.ReportRequest
```
흠 .. 이 PR과는 관련 없을수도 있지만 이렇게 repository를 모아두고 보니 보여서요
domain 모듈인데 data모듈을 알고 있는 게 좀 걸리네요
나중에 전체적으로 toDomain()확장 함수를 만들어서  domain 객체로 변환하도록 수정하면 좋겠네요

kanghwayuri commented on app/src/main/java/com/eatssu/android/domain/usecase/review/GetMealReviewListUseCase.kt:
```diff
@@ -1,17 +1,17 @@
-        reviewRepository.getReviewList(MenuType.VARIABLE.toString(), mealId, 0)
+    ): List<Review> =
+        reviewRepository.getReviewList(MenuType.VARIABLE.toString(), mealId, 0)?.toReviewList()
+            ?: emptyList()
```
근데 ReviewRepositoryImpl에서는 .orNull()로 실패 시 일부러 null로 반환되게 했는데 여기서 null 처리를 하면 .orNull()을 별도로 하는 이유가 있을까요??
어차피 UI단에서는 UseCase를 호출할 거 같은데 잘 이해가 안 돼서 여쭤봅니다!

PeraSite replied:
저도 이제 보니 일관성이 없게 작성했네요..!?
굳이 UseCase가 필요 없어 Repository 함수를 ViewModel에서 바로 호출하는 경우에만 Repository에서 도메인 변환까지 모두 수행했고,
UseCase를 쓰는 경우에는 Repository는 최대한 순수한 Response만 반환하려고 했던 것 같은데... 

모든 상황에 다 UseCase를 만드는 것보다는 Repository를 호출하는 쪽이 그래도 나은 것 같아 최대한 Repository에서 다 변환하게끔 수정하겠습니다!

kanghwayuri commented on app/src/main/java/com/eatssu/android/domain/usecase/review/GetMenuReviewListUseCase.kt:
```diff
@@ -1,17 +1,17 @@
-        reviewRepository.getReviewList(MenuType.FIXED.toString(), 0, menuId)
+    ): List<Review> =
+        reviewRepository.getReviewList(MenuType.FIXED.toString(), 0, menuId)?.toReviewList()
+            ?: emptyList()
```
여기두욤

kanghwayuri commented on app/src/main/java/com/eatssu/android/domain/usecase/review/GetMyReviewsUseCase.kt:
```diff
@@ -1,14 +1,13 @@
-    suspend operator fun invoke(): Flow<BaseResponse<MyReviewResponse>> =
-        userRepository.getUserReviews()
+    suspend operator fun invoke(): List<Review> =
+        userRepository.getUserReviews()?.toReviewList() ?: emptyList()
```
여기두

kanghwayuri commented on app/src/main/java/com/eatssu/android/di/network/ApiResultCallAdapterFactory.kt:
```diff
@@ -0,0 +1,46 @@
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+
+class ApiResultCallAdapterFactory : CallAdapter.Factory() {
```
혹시 실행해서 테스트 해보셨을까요?
로그인하면 바로 "Return 타입은 ApiResult<T> 형태여야 합니다." 오류가 나네요


```kotlin
@POST("oauths/kakao")
    suspend fun loginWithKakao(
        @Body request: LoginWithKakaoRequest,
    ): ApiResult<TokenResponse>
```

CallAdapterFactory가 Call<ApiResult<T>> 타입만 처리하도록 작성되어 있어서 suspend 함수의 반환 타입(ApiResult<Token>)을 인식하지 못하고 예외를 던지는 거 같아요 

이부분 확인한번 부탁드려요!

PeraSite replied:
loginWithKakao 에서는 아니고, 최초 IntroActivity에서 health check를 할 때 /actuator/health의 반환 데이터 형태가 BaseResponse가 아니어서 발생하던 오류는 해결했습니다.

앱을 킬 때가 아니라 카카오 로그인 한 이후에 "Return 타입은 ApiResult 형태여야 합니다." 오류가 발생하신걸까요?

kanghwayuri commented on app/src/main/res/layout/activity_server_error.xml:
```diff
@@ -0,0 +1,21 @@
+        android:background="@color/primary"
+        tools:context=".presentation.error.ServerErrorActivity">
+
+    <ImageView
```
이거 디자인 나왔나요?

PeraSite replied:
디자인이 나온 것은 아니고 임의로 IntroActivity와 같은 디자인을 사용했습니다! 어차피 팝업때문에 아무것도 안보이더라구요.

10월 12일에 진행한 87차 회의에서 Dialog나 오류 처리, 토스트 메세지 등 플랫폼 간 경험을 전반적으로 통일하는 TF를 오늘 진행하기로 해서 이 과정에서 해결해보겠습니다!

kangyuri1114 reviewed (commented):
좋은 의견 잘 봤습니다! 처음보는 구조, 코드라 질문이 좀 많습니다 ㅎㅎ.. 천천히 읽어보고 답변 부탁드려요!

>Repository에서 ApiResult를 완전히 가리고 결과값만 던져주는 구조와, ApiResult의 data만 도메인 객체로 바꾸고 반환해주는 구조 중 무엇이 더 좋을까요? 호출하는 입장에서 '어떤 오류'가 발생했는가를 상세히 알아야할 필요가 있는지에 따라 결정하면 될 것 같습니다.

그리고 질문에 대한 저의 의견은 
Repository가 ApiResult를 완전히 숨기기보다는
에러 타입만 도메인 구조로 변환해 상위 계층에 전달하는 방식은 어떨까해요
저는 사용자도 어느정도 오류의 원인을 조금 더 자세히 알면 좋을 거 같아서요

PeraSite committed:
fix: /actuator/health 가 BaseResponse 형태를 따르지 않아 발생하던 문제 수정

정상적인 Response가 왔다면 BaseResponse가 아니더라도 넘어갈 수 있게 HealthCheckService.checkHealth의 반환 타입을 Unit으로 타입을 변경했습니다.

PeraSite committed:
docs: ApiResult의 각 data class 주석 추가

PeraSite committed:
docs: ApiResult 매핑 함수 주석 추가

PeraSite committed:
chore: 기존 PartnershipRepositoryImpl 주석 복구

PeraSite committed:
refactor: ApiResult.map 외부 확장 함수로 분리

PeraSite committed:
refactor: true 반환 one-liner로 변경

PeraSite committed:
refactor: getUserNickName ApiResult 매핑 형태 개선

PeraSite committed:
refactor: 불필요한 orNull 및 emptyList() 처리 개선

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/mypage/userinfo/UserInfoViewModel.kt:
```diff
@@ -60,33 +56,31 @@ class UserInfoViewModel @Inject constructor(
+
+        val valid = validateUserNameUseCase(inputNickname)
+
+        if (!valid) {
+            _uiState.update {
+                it.copy(
+                    loading = false,
+                    error = false,
+                    isEnableName = false,
+                    toastMessage = "이미 사용 중인 닉네임 입니다.",
+                )
             }
+            return@launch
+        }
+
+        _uiState.update {
+            it.copy(
+                loading = false,
+                error = false,
+                isEnableName = true,
+                toastMessage = "사용가능한 닉네임 입니다.",
+                isNicknameChecked = true,
+            )
```
오 이거 되게 명시적인데요 👍

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/login/LoginViewModel.kt:
```diff
@@ -41,24 +41,34 @@ class LoginViewModel @Inject constructor(
 
-                _uiState.value = UiState.Success(LoginState.LoginSuccess)
+            when (val result = loginUseCase(LoginWithKakaoRequest(email, providerID))) {
+                is ApiResult.Success -> {
```
ApiResult가 어느 레이어에 있을지를 생각해보면 좋을 것 같아요

data 레이어인데 presentation 레이어에서 호출할 수는 없을 것 같아서, 이 부분을 좀더 생각해보시면 좋을 것 같숨다

정 안된다면 core에 ㅎㅎㅎ

PeraSite replied:
헉 여기는 왜 까먹고 리팩토링 안했는지 모르겠네요! UseCase 단에서 ApiResult 처리하고 UI로 넘기게 수정할게요!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/base/BaseActivity.kt:
```diff
@@ -47,6 +50,9 @@ abstract class BaseActivity<B : ViewBinding>(
     private lateinit var firebaseRemoteConfigRepository: FirebaseRemoteConfigRepository
 
+    @Inject
+    lateinit var checkServerHealthUseCase: CheckServerHealthUseCase
```
checkServerHealthUseCase를 액티비티에서 바로 호출하는게 조금 어색해보이는데 뷰모델로 개선할 수 있으면 좋을 거 같어요

그리고 Compose로 아예 넘어갔을때에 BaseActivity를 Compose용으로 새로 만들거나 아니면 BaseActivity가 아닌 다른 방법으로??? 하는 것두 같이 고민해보면 좋을 것 같아용

PeraSite replied:
제가 이전에 PR 올린 것처럼 네트워크 처리하는 Retrofit 관련 클래스에서 UI를 직접 호출하지 않으면서 서버와 통신할 수 없다는 상태를 어떻게 판별할까... 고민을 많이 해서 내린 결론이
BaseActivity에서 주기적으로 health check를 하자는 것이었거든요! 근데 다시 생각해보니 TokenEventBus처럼 글로벌하게 네트워크가 죽었다는 것을 공지할 수 있는 event observer 패턴으로 하면 될 것 같아요! 
수정하겠습니당

HI-JIN2 commented on app/src/main/java/com/eatssu/android/data/dto/response/HealthCheckResponse.kt:
이거 만들어놓고 안쓰는 것 같아염

PeraSite replied:
삭제 완료했습니다!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/data/service/HealthCheckService.kt:
```diff
@@ -0,0 +1,10 @@
+    @GET("actuator/health")
+    suspend fun checkHealth(): ApiResult<Unit>
+}
+
```
이거 혹시... 서버가 진짜 나갔을때 이게 isSuccess가 false오는지 확인해보셨을까용?? 

저번에 서버팀에서 서버 내려가면 모니터링 하던거 날라간다고? 해던 것 같아서 이 api로 헬스체크를 하는게 의미가 있는 행위인지 확인해보아야할 것 가타요

PeraSite replied:
실제로 직접 구현한 Health Check 엔드포인트가 아니라 Spring Boot Actuator에서 제공해주는 것이라 신뢰하긴 어려운게 맞습니다.
다만 HC를 하는 것의 목적이 백엔드 서버가 살아있나? HTTP 통신이 가능한 상태인가?를 확인하는 것이라 어떤 값이 오는지는 중요하지 않다고 생각했어요. 그래서 Response 타입도 Unit으로 변경한겁니다!

/actuator/health에 요청을 날렸는데 소켓 오류, TLS 오류, DNS 오류 등등 통신이 불가한 상태인지 확인하는 목적입니다!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/cafeteria/menu/MenuViewModel.kt:
```diff
@@ -16,42 +16,39 @@ import kotlinx.coroutines.flow.MutableStateFlow
 
-    private val _todayMealDataDodam = MutableStateFlow<ArrayList<GetMealResponse>>(arrayListOf())
-    val todayMealDataDodam: StateFlow<ArrayList<GetMealResponse>> = _todayMealDataDodam
+    private val _todayMealDataDodam = MutableStateFlow<List<GetMealResponse>>(emptyList())
```
ArrayList에서 List로 바꾼 이유가 있을까용?

PeraSite replied:
잇슈 안드로이드 레포지토리 전반적으로 ArrayList와 List를 혼용하는 것을 점진적으로 개선해보려고 변경했습니다. To Reviewers 섹션에 코멘트를 제가 까먹고 못 달았네요!

ArrayList는 MutableList기 때문에 정말 mutable한 list가 필요한 것이 아니라면 언제 어디서나 리스트를 실수로라도 수정할 수 있으니 사용을 지양하는 것이 좋다고 생각합니다.
Flow를 선언할 때 제네릭스에 쓰는 타입도 MutableList 인터페이스가 아니라 ArrayList라는 특정한 구현체를 명시적으로 작성하는 것이 좋지 않다고 생각했어요.
정말 극한으로 가면 List의 특정 인덱스를 가져올 필요가 없다? -> Collection or Iterable 사용이 맞긴 하지만 사용의 편의성을 위해 List를 쓰려고 합니다!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/cafeteria/menu/MenuViewModel.kt:
```diff
@@ -64,95 +61,62 @@ class MenuViewModel @Inject constructor(
+                        Restaurant.DODAM -> _todayMealDataDodam.value = restaurantMenuData
+                        Restaurant.DORMITORY -> _todayMealDataDormitory.value = restaurantMenuData
+                        Restaurant.FACULTY -> _todayMealDataFaculty.value = restaurantMenuData
+                        else -> Timber.d("onResponse 실패. 잘못된 식당입니다.")
```
요거 식단 쪽 클린 코드 작업 처음 오셨을때 부탁드린 것 같은데! 아직 멀었군여

PeraSite replied:
이것도 같이 할까말까 했는데 바로 해서 올리겠습니다 ㅎㅎ.. 저도 이 부분 보면서 와 말도 안된다!! 생각하고 있었거든요

PeraSite replied:
mutable하지 않은 훨씬 깔끔한 구조로 변경했습니다!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/base/BaseActivity.kt:
```diff
@@ -176,13 +189,37 @@ abstract class BaseActivity<B : ViewBinding>(
+                return@launch
+            }
+        }
     }
```
P1

음... BaseActivity에서 헬스체크 API 콜을 하는게 썩 좋아보이진 않아요. 호출 비용이.. 너무 많이 듭니다. ㅜㅜ

현재 구현 방식보다는, API 응답에 대한 결과로써 502 Bad Gateway 나 504 Gateway Timeout을 처리하는게 맞다 생각됩니다.

PeraSite replied:
위에 작성한 것처럼 오류가 발생했을 때 TokenEventBus처럼 Retrofit CallAdapter 단에서 처리할 수 있도록 수정하겠습니다!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/data/model/ApiResult.kt:
```diff
@@ -0,0 +1,58 @@
+}
+
+// ApiResult가 성공인지 아닌지 여부 확인
+fun ApiResult<Unit>.isSuccess(): Boolean = this is Success
```
요게.. 살짝 위험해보입니다
계속해서 개발자가 바뀌는 저희 팀 특성상 일관성을 유지하는것이 좀 어려운 과제라고 생각해요

저희 팀은 base response를 쓰고 있죠? 
응답 코드도 두개고, isSuccess도 두개에요

근데 간혹가다가, 200 속에 400을 보내는 개발자가 있을 수 있습니다

그래서 살짝 위험한 코드가 아닌지..! 조심스레 의견을 남깁니다

<img width="778" height="736" alt="image" src="https://github.com/user-attachments/assets/37f71980-ec76-40b1-9b22-370d1a10b021" />

PeraSite replied:
ㅋㅋㅋㅋ 올려주신 짤처럼 저도 HTTP Response Code만 신뢰할 수 없으니, BaseResponse라는 래퍼 정보를 기반으로 오류 처리하려고 고민을 많이 했어요.
그래서 ApiResultCall.kt의 toApiResult 함수를 보시면
1. HTTP Response Code가 2xx, 3xx대면 Fail
2. BaseResponse 형태가 아니면 Fail
3. BaseResponse.isSuccess가 false면 Fail
4. BaseResponse.data가 null이면 Fail

이렇게 BaseResponse까지 검증해서 처리하고 있어요!
그래서 ApiResult is Success라면 HTTP 반환 값뿐만 아니라 백엔드에서도 정말 진짜 성공한거가 맞습니다!

HI-JIN2 reviewed (commented):
리뷰 늦게 드려서 죄송합니다 ㅜㅜㅜ  넘 수고 많으셨어요

몇가지 코멘트를 남겨두었으니 확인 부탁드립니다 🙇‍♀️

PeraSite committed:
refactor: ApiResult가 UI 레이어로 넘어가지 않게 loginWithKakao 수정

PeraSite committed:
refactor: 미사용 Response 객체 삭제

PeraSite committed:
refactor: 하드 코딩된 부분을 Restraunt enum 사용하게 수정

PeraSite committed:
refactor: 안쓰는 함수 제거

PeraSite committed:
refactor: 어떤 식당 메뉴 가져올 것인지 한번에 호출하게 변경

PeraSite committed:
refactor: 불필요한 UiState.Success의 nullable 제거

PeraSite committed:
fix: UiState.Success 변경 대응

PeraSite committed:
feat: Health Check 관련 코드 삭제

PeraSite committed:
feat: NetworkErrorEventBus로 네트워크 오류 발생 시 ServerErrorActivity로 이동

PeraSite committed:
refactor: Indent 적게 수정

PeraSite committed:
refactor: Deprecate된 arguments.getSerializable 개선

PeraSite committed:
chore: 불필요한 out 삭제

PeraSite committed:
chore: 불필요한 context 삭제

PeraSite committed:
fix: Deprecate된 fragmentManager 접근 수정

HI-JIN2 commented on app/src/main/java/com/eatssu/android/data/repository/PartnershipRepositoryImpl.kt:
```diff
@@ -14,23 +17,15 @@ class PartnershipRepositoryImpl @Inject constructor(
 
-    // 유저의 학과에 해당하는 제휴 정보 조회
-    override suspend fun getUserCollegePartnerships(): List<Partnership> {
-        return userService.getUserDepartmentPartnerships()
-            .result
-            ?.map { it.toDomain() } ?: emptyList()
-    }
+    override suspend fun getUserCollegePartnerships(): List<Partnership> =
+        userService.getUserDepartmentPartnerships().map { list -> list.map { it.toDomain() } }
+            .orEmptyList()
```
기존 주석 날라간 것 같아유

PeraSite replied:
추가하겠습니다!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/intro/IntroActivity.kt:
```diff
@@ -49,13 +56,26 @@ class IntroActivity : AppCompatActivity() {
+                            }
+                        startActivity(intent)
+                        finish()
+                    }
```
지금 도메인을 바꾸지 않은채로 빌드를 해보면, UnknownHostException이 뜨죠. 
intro 에서 UiEvent.NavigateToServerError으로 떨궈주지도 않고 baseActivity를 상속하는것도 아니어서 
intro -> login 화면으로 갑니다 (intro > error 화면으로 가는게 적합하겟쬬?)

문제점을 해결해주지 못하고 있어요.. 요부분 개선이 필요할 것 같아요

PeraSite replied:
이 부분 캐치 못했네요 감사합니다!
UiEvent.NavigateToServerError 자체가 필요하지 않은 것 같아서 삭제하고, IntroActivity에서 NetworkErrorEventBus를 쓰게끔 수정할게요!

HI-JIN2 commented on app/src/main/java/com/eatssu/android/presentation/cafeteria/menu/MenuViewModel.kt:
```diff
@@ -2,158 +2,81 @@ package com.eatssu.android.presentation.cafeteria.menu
+                async {
+                    when (restaurant.menuType) {
+                        MenuType.FIXED -> {
+                            when (val result = menuService.getFixMenu(restaurant.toString())) {
```
옴... 제가 말한 클린아키텍쳐 리팩토링은 뷰모델에서 바로 서비스를 호출하는걸 레이어를 나눠서 repository > usecase로 나누어달라는 거였는데..!

이번 PR에서 작업해야할 부분은 아닌것 같아서 넘길게유

PeraSite replied:
클린 코드만 생각하다보니 구조만 갈아엎고 service 호출 분리를 깜빡했네요!
이미 건드려서 커밋이 올라간 이상 어차피 하긴 해야하니 같이 작업해서 올리겠습니다!

HI-JIN2 reviewed (commented):
동작 관련해서 체크해보아야할게 있어보여서 코멘트 달았습니다!!

PeraSite committed:
docs: 누락된 주석 추가

PeraSite committed:
refactor: BaseActivity, IntroActivity에서 공통적으로 NetworkErrorEventBus에 구독하도록 수정

PeraSite committed:
Merge branch 'develop' into refactor/error-handling

PeraSite committed:
feat: HealthCheck 관련 Service, Repository, UseCase 추가

PeraSite committed:
feat: IntroViewModel에서 최초 실행 시 health check하게끔 수정

PeraSite committed:
refactor: MenuViewModel 클린 아키텍처 적용

PeraSite committed:
chore: 오타 수정

kangyuri1114 reviewed (approved):
Repository가 ApiResult를 리턴하는 구조로 모두 변경되면서 viewmodel 단의 runcatching도 웬만하면 필요가 없어졌네요! 이부분도 차차 필요없어진 부분은 걷어내면 좋을 것 같습니다

PeraSite committed:
refactor: MapViewModel runCatching 리팩토링

PeraSite committed:
refactor: UserInfoViewModel try-catch 리팩토링

PeraSite committed:
refactor: MyPageViewModel runCatching 리팩토링

PeraSite committed:
refactor: SignOutViewModel try-catch 리팩토링

PeraSite committed:
refactor: MainViewModel runCatching 리팩토링
